client/gutenberg/editor/calypsoify-iframe.jsx

```diff
@@ -5,40 +5,136 @@
  */
 import React, { Component, Fragment } from 'react';
 import { connect } from 'react-redux';
-import { pickBy } from 'lodash';
+import { map, pickBy } from 'lodash';
 
 /**
  * Internal dependencies
  */
+import MediaLibrarySelectedData from 'components/data/media-library-selected-data';
+import MediaModal from 'post-editor/media-modal';
+import MediaActions from 'lib/media/actions';
 import { getSelectedSiteId } from 'state/ui/selectors';
-import { getSiteOption, getSiteAdminUrl } from 'state/sites/selectors';
+import { getSiteOption, getSiteAdminUrl, getSiteSlug } from 'state/sites/selectors';
 import { addQueryArgs } from 'lib/route';
+import {
+	getEnabledFilters,
+	getDisabledDataSources,
+	mediaCalypsoToGutenberg,
+} from './hooks/components/media-upload/utils';
 
 /**
  * Style dependencies
  */
 import './style.scss';
 
 class CalypsoifyIframe extends Component {
+	state = {
+		isMediaModalVisible: false,
+	};
+
+	constructor( props ) {
+		super( props );
+		this.iframeRef = React.createRef();
+	}
+
+	componentDidMount() {
+		window.addEventListener( 'message', this.onMessage, false );
+	}
+
+	componentWillUnmount() {
+		window.removeEventListener( 'message', this.onMessage, false );
+	}
+
+	onMessage = ( { data, origin } ) => {
+		if ( ! data || origin.indexOf( this.props.siteSlug ) < 0 ) {
+			return;
+		}
+
+		const { action, type } = data;
+
+		if ( 'gutenbergIframeMessage' !== type ) {
+			return;
+		}
+
+		if ( 'loaded' === action ) {
+			const { port1: portToIframe, port2: portForIframe } = new MessageChannel();
+
+			this.iframePort = portToIframe;
+			this.iframePort.addEventListener( 'message', this.oniFramePortMessage, false );
+			this.iframePort.start();
+
+			this.iframeRef.current.contentWindow.postMessage( { action: 'initPort' }, '*', [
+				portForIframe,
+			] );
+		}
```

Comment:
I'd love to see us detach the window listener here so it doesn't dangle around. I thought that we might have to introduce a new instance property on the class to prevent double-removing it on unmount but according to the [MDN guide for `removeEventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener#Notes) that should be a benign operation.

that means here after establishing the message port we might want to add…

```js
window.removeEventListener( 'message', this.onMessage, false );
```